WIP - Raise exception when disk object corruption detected

ardumont opened the pull request:
I still have implementation questions pending in the related issue.

I prefer opening the PR now to show the current state so that we can
discuss freely the implementation.

Related #582

(rebased as i was not up-to-date with upstream/master)

jelmer commented:
Having check report these and not raise an exception seems reasonable to me. Similarly, when iterating over all objects in the repo, it seems reasonable to warn and skip problematic objects.

However, I don't think we should silently ignore corrupted files, that just results in more confusion. If the user requests an object that is unparseable, that should just raise an exception like it did before (Perhaps a slightly different exception than ValueError though).

That would also be more in line with C git.

ardumont commented:
I did not reply as i was trying to improve the current patch and go in the direction you proposed.
I'm not feeling that i did good though.

> Having check report these and not raise an exception seems reasonable to me.

Ok. 

I don't see where i can do that check in the objects though (since those are not path format).
What am i missing?

My understanding must be too much tied with the corruption case i have in #582 (empty file).
The only corruption i see is that of the empty file (which seems to be only possible in the loose objects pool).

>  Similarly, when iterating over **all** objects in the repo, it seems reasonable to warn and skip problematic objects.

Ok. I don't see where i can treat this error for all objects now.
I'm plugged on the loose objects since i only see path objects manipulated there.

> However, I don't think we should silently ignore corrupted files, that just results in more confusion. 

I agree with that.

It's just that i did not see clearly where i could have put this log behavior :)
So, i finally just added log to the DiskObjectStore and log from there.
I'm not sure that that's what you are suggesting though.

> If the user requests an object that is unparseable, that should just raise an exception like it did before (Perhaps a slightly different exception than ValueError though).

I did go with a new CorruptedFormatException inheriting from dulwich.errors.FileFormatException.

Looking forward to hear your thoughts on this.

Thanks for your help.

Cheers,

jelmer commented on dulwich/errors.py:
```diff
@@ -164,6 +164,10 @@ class ObjectFormatException(FileFormatException):
 
 
+class CorruptedFormatException(FileFormatException):
+    """Indicates a corruption in the object's disk representation."""
```
Any FileFormatException is a sign of an invalid disk representation; perhaps you mean something like EmpyFileException?

jelmer commented on dulwich/object_store.py:
```diff
@@ -557,14 +560,19 @@ def _iter_loose_objects(self):
-            if os.lstat(path).st_size == 0:
-                return None
             return ShaFile.from_path(path)
+        except CorruptedFormatException:
```
You'd probably want to let this bubble up to the caller. If they explicitly asked for this object (rather than e.g. as part of iterating all the objects in the repo), they'd want to know about corruption.

jelmer commented on dulwich/tests/test_object_store.py:
```diff
@@ -346,13 +346,19 @@ def test_alternates(self):
         self.store.add_object(testobject)
-        self.assertEqual((Blob.type_num, b'yummy data'),
-                         self.store.get_raw(testobject.id))
+        assert ((Blob.type_num, b'yummy data') ==
```
Please use standard unittest mehotds; e.g. self.assertXXX rather than assert.

jelmer commented on dulwich/objects.py:
```diff
@@ -383,8 +384,23 @@ def _serialize(self):
         raise NotImplementedError(self._serialize)
 
     @classmethod
+    def _check_is_corrupted(cls, path):
```
for performance reasons, please raise exceptions about formatting issues where they are noticed, rather than checking for all corruption up front. E.g. in ShaFile._parse_file, raise an exception if the file was empty.

ardumont commented on dulwich/tests/test_object_store.py:
```diff
@@ -343,6 +343,23 @@ def test_alternates(self):
+                self.store.get_raw(testobject.id))
+
+        assert self.store.contains_loose(testobject.id) is True
+        assert self.store._get_loose_object(testobject.id) is not None
```
> Please use standard unittest mehotds; e.g. self.assertXXX rather than assert.

(I don't see that comment anymore but still, explaining my thought here.)

Those are state assertions needed for the setup phase of the test.
If i'm wrong here, i want a different error message to raise than the standard `self.assertXXX`'s failure message below.

jelmer replied:
I think you do want the same kinds of assertions, e.g. self.assertEqual, self.assertIs. The downsides of assert are that it doesn't return the values on either side of the assert statement, and it gets ignored if you run Python with optimizations enabled.

ardumont replied:
Ok. Will adapt accordingly then.

ardumont commented on dulwich/object_store.py:
```diff
@@ -557,12 +560,19 @@ def _iter_loose_objects(self):
+        except CorruptedFormatException:
+            self.log.warn('Object with id %s is corrupted! Skipping.' % (
+                sha.decode('ascii'), ))
+            return None
```
> You'd probably want to let this bubble up to the caller. If they explicitly asked for this object (rather than e.g. as part of iterating all the objects in the repo), they'd want to know about corruption.

(Same here, i don't see that comment anymore.)

It's the same idea you already explained to me, but now i think, i fully understand.

ardumont commented:
(I'll rebase and squash commits when we deem this ready)

jelmer commented:
Thanks, looks good to me apart from the asserts.

ardumont commented:
Rebased and squashed commits.

I kept the one fix docstring commit aside.
This is not the same perimeter so i kept them separated (for ease of potential future revert).

Tell me if you prefer i squash everything or if you prefer another PR (i was lazy on that and thought it was a bit much...).

Cheers,
